orchestrator/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,481 @@
+	defer func() {
+		err = errors.Wrapf(err, "fail to create replica %v for %v",
+			req.InstanceName, req.VolumeName)
+	}()
+
+	if err := orchestrator.ValidateRequestCreateReplica(req); err != nil {
+		return nil, err
+	}
+
+	if req.NodeID != k.currentNode.ID {
+		return nil, fmt.Errorf("incorrect node, requested %v, current %v", req.NodeID,
+			k.currentNode.ID)
+	}
+
+	cmd := []string{
+		"launch", "replica",
+		"--listen", "0.0.0.0:9502",
+		"--size", strconv.FormatInt(req.VolumeSize, 10),
+	}
+	if req.RestoreFrom != "" && req.RestoreName != "" {
+		cmd = append(cmd, "--restore-from", req.RestoreFrom, "--restore-name", req.RestoreName)
+	}
+	cmd = append(cmd, "/volume")
+
+	podNode, err := k.getCurrentNodePod()
+
+	if err != nil {
+		logrus.Errorf("fail to get daemonset pod info")
+		return nil, err
+	}
+
+	privilege := true
+	pod := &apiv1.Pod{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name: req.InstanceName,
+		},
+		Spec: apiv1.PodSpec{
+			NodeSelector: map[string]string{
+				"kubernetes.io/hostname": podNode.Spec.NodeName,
+			},
+			Containers: []apiv1.Container{
+				{
+					Name:    req.InstanceName,
+					Image:   k.EngineImage,
+					Command: cmd,
+					SecurityContext: &apiv1.SecurityContext{
+						Privileged: &privilege,
+					},
+					VolumeMounts: []apiv1.VolumeMount{
+						{
+							Name:      "volume",
+							MountPath: "/volume",
+						},
+					},
+				},
+			},
+			Volumes: []apiv1.Volume{
+				{
+					Name: "volume",
+					VolumeSource: apiv1.VolumeSource{
+						HostPath: &apiv1.HostPathVolumeSource{
+							Path: "/var/longhorn/replica/" + req.InstanceName,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	podReplica, err := k.cli.CoreV1().Pods(apiv1.NamespaceDefault).Create(pod)
+
+	if err != nil {
+		return nil, err
+	}
+
+	req.InstanceID = podReplica.ObjectMeta.Name
+
+	defer func() {
+		if err != nil {
+			k.StopInstance(req)
+			k.DeleteInstance(req)
+		}
+	}()
+
+	instance, err = k.InspectInstance(req)
+	if err != nil {
+		logrus.Errorf("fail to inspect when create replica %v of %v, cleaning up: %v", req.InstanceName, req.VolumeName, err)
+		return nil, err
+	}
+
+	timeout := WaitAPITimeout
+	// More time for backup restore, may need to customerize it
+	if req.RestoreFrom != "" && req.RestoreName != "" {
+		timeout = timeout * 10
+	}
+	url := "http://" + instance.IP + ":9502/v1"
+
+	//this fuction maybe a risk, when the IP is unreachable, it will block forever
+	if err := util.WaitForAPI(url, timeout); err != nil {
+		return nil, err
+	}
+
+	return instance, nil
+}
+
+func (k *Kubernetes) waitForPodReady(podName string) (*apiv1.Pod, error) {
+	c := time.After(time.Second * time.Duration(WaitPodTimeout))
+
+	for i := 0; i < WaitPodCounter; i++ {
+		select {
+		case <-c:
+			pod, err := k.cli.CoreV1().Pods(apiv1.NamespaceDefault).Get(podName, meta_v1.GetOptions{})
+			if err != nil {
+				return nil, fmt.Errorf("get pod fail %v", podName)
+			}
+			if pod.Status.PodIP == "" {
+				c = time.After(time.Second)
```

Comment:
Essentially you're waiting for 3 seconds for the IP, which is very aggressive as well. Just use `WaitPodTimeout` here as well.